src/abstractdataframe/join.jl

```diff
@@ -91,6 +91,346 @@ _rename_cols(old_names::AbstractVector{Symbol},
            (renamecols isa Function ? Symbol(renamecols(string(n))) : Symbol(n, renamecols))
            for n in old_names]
 
+prepare_on_col() = throw(ArgumentError("at least one on column required when joining"))
+prepare_on_col(c::AbstractVector) = c
+prepare_on_col(cs::AbstractVector...) = tuple.(cs...)
+
+# Return if it is allowed to use refpool instead of the original array for joining.
+# There are multiple conditions that must be met to allow for this.
+# In particular we must be able to find a sentinel value that will be used
+# in mapping to signal that the mapping to the longer column is not present
+# in some rare cases it is impossible to find such a sentinel (e.g. for
+# CategoricalArray with missing and having number of levels exactly equal
+# to the typemax of element type of ref pool)
+function check_mapping_allowed(short, long)
```

Comment:
@nalimilan - no rush with reviewing this (we first need to tag CategoricalArrays.jl to have a good test coverage), but here (and I will indicate also below) are critical parts of the DataAPI.jl related code logic. Thank you!

Reply:
Would it be possible to use `nothing` as a sentinel? That would probably be simpler, and maybe not slower?

Reply:
It would if DataAPI.jl guaranteed that `nothing` is not allowed as a valid ref, which it does not guarantee now. Also it will be a bit faster to use a sentinel that is integer (less memory allocation and faster processing probably around 10% faster).